File: Semester 4/Programming and Projection Enironments/Proiect_MPP/src/main/java/root/proiect_mpp/repositories/AbstractRepository.java
package root.proiect_mpp.repositories;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import root.proiect_mpp.domain.Entity;
import root.proiect_mpp.utils.JdbcUtils;

import java.sql.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Properties;

public abstract class AbstractRepository<T extends Entity<ID>, ID> implements Repository<T, ID> {
    private final JdbcUtils dbUtils;
    protected String sqlCommand;
    protected static final Logger logger = LogManager.getLogger();

    // Class constructors //

    public AbstractRepository(Properties properties) {
        logger.info("Initializing Abstract Repository with properties: {} ", properties);
        dbUtils = new JdbcUtils(properties);
        this.sqlCommand = "";
    }

    protected abstract T extractEntity(ResultSet resultSet) throws SQLException;

    protected abstract PreparedStatement createInsertStatement(Connection connection, T entity) throws SQLException;

    protected abstract PreparedStatement createUpdateStatement(Connection connection, T entity, ID id) throws SQLException;

    protected abstract PreparedStatement createDeleteStatement(Connection connection, T entity) throws SQLException;

    protected PreparedStatement createStatementFromEntity(Connection connection, T entity) throws SQLException {
        char command = sqlCommand.charAt(0);
        return switch (command) {
            case 'I' -> createInsertStatement(connection, entity);
            case 'D' -> createDeleteStatement(connection, entity);
            default -> null;
        };
    }

    private void executeCommand(T entity) {
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement statement = createStatementFromEntity(connection, entity)
        ) {
            int result = statement.executeUpdate();
            logger.trace("Modified {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
    }

    private void executeCommand(T entity, ID id) {
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement statement = createUpdateStatement(connection, entity, id)
        ) {
            int result = statement.executeUpdate();
            logger.trace("Modified {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
    }

    private Collection<T> executeQuery() {
        List<T> elements = new ArrayList<>();
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement statement = connection.prepareStatement(sqlCommand);
             ResultSet resultSet = statement.executeQuery()
        ) {
            while (resultSet.next()) {
                elements.add(extractEntity(resultSet));
            }
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
        logger.traceExit();
        return elements;
    }

    @Override
    public void add(T elem) {
        logger.traceEntry("Saving entity {}", elem);
        executeCommand(elem);
        logger.traceExit();
    }

    @Override
    public void delete(T elem) {
        logger.traceEntry("Deleting entity {}", elem);
        executeCommand(elem);
        logger.traceExit();
    }

    @Override
    public void update(T elem, ID id) {
        logger.traceEntry("Updating entity {}", elem);
        executeCommand(elem, id);
        logger.traceExit();
    }

    public T getOne() {
        logger.traceEntry();
        Collection<T> result = executeQuery();
        if (result.isEmpty())
            return null;
        return executeQuery().iterator().next();
    }

    @Override
    public Collection<T> getAll() {
        logger.traceEntry();
        return executeQuery();
    }
}

File: Semester 4/Programming and Projection Enironments/mpp-proiect-csharp/Proiect_MPP/persistence/AbstractRepository.cs
using System.Data.SQLite;
using model;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using log4net;

namespace persistence
{
    public abstract class AbstractRepository<T, Tid> : Repository<T, Tid> where T : Entity<Tid>
    {
        protected IDictionary<string, string> properties;
        protected IDbConnection? connection;
        protected IDbCommand? sqlCommand;
        protected static readonly ILog logger = LogManager.GetLogger("AbstractRepository");

        // Class constructors //

        public AbstractRepository(IDictionary<string, string> properties)
        {
            logger.Info("Initializing Abstract Repository.");
            this.properties = properties;
            this.connection = null;
            this.sqlCommand = null;
        }

        protected abstract T extractEntity(IDataReader dataReader);

        private void executeCommand()
        {
            try
            {
                int result = sqlCommand.ExecuteNonQuery();
                logger.InfoFormat("Modified {0} instances", result);
            }
            catch (Exception ex)
            {
                logger.Error(ex);
            }
        }

        private List<T> executeQuery()
        {
            List<T> elements = new List<T> { };
            try
            {
                using (var dataReader = sqlCommand.ExecuteReader()) 
                {
                    while (dataReader.Read())
                    {
                        elements.Add(extractEntity(dataReader));
                    }
                }
            }
            catch (Exception ex)
            {
                logger.Error(ex);
            }
            return elements;
        }

        public virtual int add(T item)
        {
            logger.InfoFormat("Saving entity {0}", item);
            executeCommand();
            return 0;
        }

        public virtual void delete(T item)
        {
            logger.InfoFormat("Deleting entity {0}", item);
            executeCommand();
        }

        public virtual void update(T item, Tid id)
        {
            logger.InfoFormat("Updating entity {0}", item);
            executeCommand();
        }
        public virtual T? findById(Tid id)
        {
            logger.InfoFormat("Finding entity after id {0}", id);
            return getOne();
        }

        public virtual List<T> findAll()
        {
            logger.Info("Finding all elements");
            return executeQuery();
        }

        public T? getOne()
        {
            List<T> elements = executeQuery();
            if (elements.Count() == 0)
            {
                return null;
            }
            return elements[0];
        }
    }
}
